Bio/PDB/internal_coords.py

```diff
@@ -0,0 +1,2765 @@
+
+        to get only backbone atoms plus amide proton, use:
+
+        `IC_Residue.accept_atoms = IC_Residue.accept_mainchain + ('H',)`
+
+        to convert D atoms to H, set `AtomKey.d2h = True` and use:
+
+        `IC_Residue.accept_atoms = accept_mainchain + accept_hydrogens + accept_deuteriums`
+
+        There is currently no option to output internal coordinates with D
+        instead of H
+
+    accept_resnames: tuple
+        list of 3-letter residue names for HETATMs to accept when generating
+        internal coordinates from atoms.  HETATM sidechain will be ignored, but normal
+        backbone atoms (N, CA, C, O, CB) will be included.  Currently only
+        CYG, YCM and UNK; override at your own risk.  To generate
+        sidechain, add appropriate entries to ic_data_sidechains in
+        ic_data.py and support in dihedra_from_atoms()
+    gly_Cbeta: bool default False
+        override class variable to True to generate internal coordinates for
+        glycine CB atoms in dihedra_from_atoms().
+
+        `IC_Residue.gly_Cbeta = True`
+
+    scale: optional float
+        used for OpenSCAD output to generate gly_Cbeta bond length
+
+    Parameters (__init__)
+    ---------------------
+    parent: biopython Residue object this class extends
+    NO_ALTLOC: bool default False
+    Disable processing of ALTLOC atoms if True, use only selected atoms.
+
+
+    Methods
+    -------
+    applyMtx()
+        multiply all IC_Residue atom_cords by passed matrix
+    assemble(atomCoordsIn, transforms)
+        Compute atom coordinates for this residue from internal coordinates
+    atm241(coord)
+        Convert 1x3 cartesian coords to 4x1 homogeneous coords
+    coords_to_residue()
+        Convert homogeneous atom_coords to Biopython cartesian Atom coords
+    dihedra_from_atoms()
+        Create hedra and dihedra for atom coordinates
+    get_angle()
+        Return angle for passed key
+    get_length()
+        Return bond length for specified pair
+    link_dihedra()
+        Link dihedra to this residue, form id3_dh_index
+    load_PIC(edron)
+        Process parsed (di-/h-)edron data from PIC file
+    pick_angle()
+        Find Hedron or Dihedron for passed key
+    pick_length()
+        Find hedra for passed AtomKey pair
+    rak(atom info)
+        Residue AtomKey - per residue AtomKey result cache
+    render_dihedra()
+        Call init_pos for each dihedron in dihedra
+    render_hedra()
+        Call init_pos for each hedron in hedra
+    set_angle()
+        Set angle for passed key (no position updates)
+    set_length()
+        Set bond length in all relevant hedra for specified pair
+    write_PIC(pdbid, chainId, s)
+        Generate PIC format strings for this residue
+
+    """
+
+    # add 3-letter residue name here for non-standard residues with
+    # normal backbone.  CYG for test case 4LGY (1305 residue contiguous
+    # chain)
+    accept_resnames = ("CYG", "YCM", "UNK")
```

Comment:
Maybe have this as a parameter the user can pass, otherwise throw an error or ignore unknown residues?

Reply:
This one I have not addressed, easy to do but want to understand more about the reasoning.  

I wanted it somewhat hard to access without thought, as plugging any HETATM is likely to break things, and what's *really* needed for a new residue is to go and edit the ic_data tables to put in a set of di/hedra to build the thing.

I agree the list is pretty arbitrary - UNK seems reasonable because I think that will just be backbone but could not resolve the sidechain (and for some reason the sequence is unknown?), while CYG and YCM are ones I made the effort to check fit into the peptide backbone and they helped me get bigger proteins going through the cullpdb lists.  Clearly though I did not make the effort to go and work out the angle sets.

Without the entry you just get a chain break and extra noise about atoms that don't match doing the rebuild test.  See your ACE/NME caps with current code.